### PR TITLE
Managed Iceberg hive support and integration tests

### LINKED ISSUE #32050: [Bug]: IcebergIO opens a writer using the table's schema, which can cause data loss
### What happened?

I'm experimenting with HiveCatalog and noticed that when writing data with nested records, there is data loss. Specifically, the nested records are never committed to the table. I read back null values instead. I can confirm this happens even without our Beam library:

```java
    TableIdentifier tableIdentifier =
        TableIdentifier.parse(String.format("%s.%s", TEST_DB, TEST_TABLE));
    org.apache.iceberg.Schema icebergSchema = IcebergUtils.beamSchemaToIcebergSchema(ROW_SCHEMA);
    Table table = catalog.createTable(tableIdentifier, icebergSchema);

    // write
    List<Record> records = LongStream.range(1, 10).boxed().map(l -> IcebergUtils.beamRowToIcebergRecord(icebergSchema, ROW_FUNC.apply(l))).collect(Collectors.toList());
    String filepath = table.location() + "/" + UUID.randomUUID();
    OutputFile file = table.io().newOutputFile(filepath);
    DataWriter<Record> writer =
            Parquet.writeData(file)
//                    .schema(table.schema())    <---- xxxx this is the problematic line xxxx
                    .schema(icebergSchema)
                    .createWriterFunc(GenericParquetWriter::buildWriter)
                    .overwrite()
                    .withSpec(table.spec())
                    .build();
    for (Record rec: records) {
      System.out.println("xxx writing: " + rec);
      writer.write(rec);
    }
    writer.close();
    AppendFiles appendFiles = table.newAppend();
    String manifestFilename = FileFormat.AVRO.addExtension(filepath + ".manifest");
    OutputFile outputFile = table.io().newOutputFile(manifestFilename);
    ManifestWriter<DataFile> manifestWriter;
    try (ManifestWriter<DataFile> openWriter = ManifestFiles.write(table.spec(), outputFile)) {
      openWriter.add(writer.toDataFile());
      manifestWriter = openWriter;
    }
    appendFiles.appendManifest(manifestWriter.toManifestFile());
    appendFiles.commit();

    
    // read
    table = catalog.loadTable(tableIdentifier);
    TableScan tableScan = table.newScan().project(icebergSchema);
    for (CombinedScanTask task : tableScan.planTasks()) {
      InputFilesDecryptor decryptor = new InputFilesDecryptor(task, table.io(), table.encryption());
      for (FileScanTask fileTask : task.files()) {
        InputFile inputFile = decryptor.getInputFile(fileTask);
        CloseableIterable<Record> iterable =
                Parquet.read(inputFile)
                        .split(fileTask.start(), fileTask.length())
                        .project(icebergSchema)
                        .createReaderFunc(
                                fileSchema -> GenericParquetReaders.buildReader(icebergSchema, fileSchema))
                        .filter(fileTask.residual())
                        .build();

        for (Record rec : iterable) {
          System.out.println("xxx reading: " + rec);
        }
      }
    }

```

I've tried the same with HadoopCatalog and everything seems to be working fine. I'm not sure why I'm seeing this error with HiveCatalog. There may be a gimmick in how the catalog fetches and returns the table schema. 

I believe in this case we shouldn't rely on the catalog, and instead create our writers using the schema of the records in our PCollection. ie. line 70 here: 
https://github.com/apache/beam/blob/21009e68eabc9d8e05c5cfadaf5016fa23646fb7/sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/RecordWriter.java#L66-L70

### Issue Priority

Priority: 1 (data loss / total loss of function)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by ahmedabu98**
This PR does the following to allow support for HiveCatalog with Managed Iceberg:
- Expose option to pass through Hadoop Configuration properties (this was needed specifically for Hive, but others will benefit from this too)
- Determine what extra dependencies are needed to use HiveCatalog
  - Provide a shaded jar that includes these dependencies
- Add integration tests for reading and writing using HiveCatalog
  - Spin up a local Hive metastore and use a GCS bucket as a warehouse
- Fix bugs in Iceberg conversion utils

Note that the code in `sdk/io/iceberg/hive/testutils/` (60~70% of this PR) is mostly copied from Iceberg's integration test directory: https://github.com/apache/iceberg/tree/main/hive-metastore/src/test/java/org/apache/iceberg/hive.

### CODE REVIEW

**Comment by ahmedabu98** [comment]
assign set of reviewers

**Comment by ahmedabu98** [comment]
Blocked on #32095

**Comment by ahmedabu98** [comment]
#32095 has been merged and this PR was just rebased to HEAD

File: sdks/java/io/iceberg/hive/src/test/resources/hive-schema-3.1.0.derby.sql [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+--
+-- Timestamp: 2011-09-22 15:32:02.024
+-- Source database is: /home/carl/Work/repos/hive1/metastore/scripts/upgrade/derby/mdb
+-- Connection URL is: jdbc:derby:/home/carl/Work/repos/hive1/metastore/scripts/upgrade/derby/mdb
+-- Specified schema is: APP
+-- appendLogs: false
+
+-- ----------------------------------------------
+-- DDL Statements for functions
+-- ----------------------------------------------
```
Comment (damccorm) [inline sdks/java/io/iceberg/hive/src/test/resources/hive-schema-3.1.0.derby.sql:30]: Do we actually need this whole file? Or can we use a subset for our use case?

**Review by damccorm — COMMENTED** [review]
Mostly LGTM, just had a minor comment.

FWIW, the size made it harder to review, this might have benefited from being split into 2 prs

File: sdks/java/io/iceberg/hive/src/test/resources/hive-schema-3.1.0.derby.sql (hunk 1 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/hive/src/test/resources/hive-schema-3.1.0.derby.sql:30]: Thanks for the comment. I was able to cut down a little more than 450 lines with some guess and checking

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**